core/src/consensus/tendermint/mod.rs

```diff
@@ -1692,7 +1688,11 @@ impl NetworkExtension for TendermintExtension {
 
 impl TimeoutHandler for TendermintExtension {
     fn on_timeout(&self, token: TimerToken) {
-        debug_assert!(token >= ENGINE_TIMEOUT_TOKEN);
+        debug_assert!(token >= ENGINE_TIMEOUT_TOKEN_NONCE_BASE);
+        if self.is_expired_timeout_token(token) {
```

Comment:
We should check the token after acquiring the tendermint lock.

Here is an example.
1. Thread A acquired the lock.
2. Timer thread B calls `on_timeout` and waiting lock.
3. Thread A changes the `step` and canceled timer and unlock.
4. Thread B acquires the lock and calls the Tendermint's `timeout` (But it should not be called)